src/Dialect/Mlir/DialectBuilder.cpp

```diff
@@ -942,4 +953,47 @@ FlatSymbolRefAttr LLVMBuilder::getOrInsertSymbolRef(ModuleOp module,
   return SymbolRefAttr::get(b.getContext(), funcName);
 }
 
+void LLVMBuilder::ifThenElse(
+    mlir::function_ref<Value(LLVMBuilder &createLLVM)> cond,
+    mlir::function_ref<void(LLVMBuilder &createLLVM)> thenFn,
+    mlir::function_ref<void(LLVMBuilder &createLLVM)> elseFn) const {
+  LLVMBuilder createLLVM(b, loc);
+
+  // Split the current block into IF, THEN, ELSE and END blocks.
+  Block *ifBlock, *thenBlock, *elseBlock, *endBlock;
+  ifBlock = b.getInsertionBlock();
+  thenBlock = ifBlock->splitBlock(b.getInsertionPoint());
+  elseBlock = b.createBlock(
+      thenBlock->getParent(), std::next(Region::iterator(thenBlock)));
+  if (elseFn)
+    endBlock = b.createBlock(
+        elseBlock->getParent(), std::next(Region::iterator(elseBlock)));
+  else
+    endBlock = elseBlock;
+
+  // Emit code for the IF block.
+  b.setInsertionPointToEnd(ifBlock);
+  Value condVal = cond(createLLVM);
+
+  // Branch the block into the THEN and ELSE blocks.
+  condBr(condVal, thenBlock, {}, elseBlock, {});
+
+  // Emit code for the THEN block.
+  b.setInsertionPointToStart(thenBlock);
+  thenFn(createLLVM);
```

Comment:
You should create a new builder with the local builder, see what is done in SCF
```C++
          SCFBuilder scfBuilder(childBuilder, childLoc);
```
It was the recommendation of MLIR to build a local builder that is specific to the loc.

Reply:
SCF is a bit different here. We are using the builder of `scf.IfOp` that uses `function_ref`, e.g. `function_ref<void(OpBuilder &, Location)> thenBuilder`, so we need to use the local location. But, if we use another builder of `scf.IfOp` that does not use `function_ref` but `setInsertionPointToStart `, the local location is not necessary, e.g. https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SparseTensor/Transforms/Sparsification.cpp#L746.

IfThenElse here for LLVM is similar to the later one, where we just emit a sequence of ops to represent ifThenElse via LLVM::CondBrOp. LLVM dialect does not have ifOp.